### PR TITLE
Further tweaks on the API.

### PR DESCRIPTION
**Pull request by lrhn**
Will also update to allow nested packages before committing.

### CODE REVIEW

File: lib/src/package_config_impl.dart [hunk 1]
```diff
@@ -179,3 +193,107 @@ void _validatePackageData(
...
+      }
+    }
+    return null;
+  }
+
+  /// Returns the [PackageTree] of the children of [package].
+  ///
+  /// Ensures that the object is allocated if necessary.
+  MutablePackageTree _treeOf(SimplePackage package) =>
+      (_packageChildren ??= {})[package.name] ??= MutablePackageTree();
```
Comment (jonasfj) [inline lib/src/package_config_impl.dart]: nit: would it hurt to write this in more than one line? :)

File: lib/src/package_config_impl.dart [hunk 2]
```diff
@@ -179,3 +193,107 @@ void _validatePackageData(
...
+/// (represented by their name) to their immediately nested packages' names.
+class MutablePackageTree implements PackageTree {
+  final List<SimplePackage> packages = [];
+  Map<String, MutablePackageTree /*?*/ > /*?*/ _packageChildren;
+
+  /// Tries to (add) `package` to the tree.
+  ///
+  /// If another package is found with the *same* path, adding fails
+  /// and that package is returned. Returns `null` on success.
+  SimplePackage /*?*/ add(int start, SimplePackage package) {
```
Comment (jonasfj) [inline lib/src/package_config_impl.dart]: nit: Maybe rename this `addOrReturnConflictingPackage(SimplePackage package, [int start = 0])`... One could also throw in the case of conflicts, as speed is unimportant in the case of a conflict.

**Review by jonasfj — APPROVED** [review]
LGTM, please feel free to ignore my nits -- I suspect you're prioritizing speed over readability :)

File: lib/src/package_config_impl.dart (hunk 1 above)
Comment (lrhn) [reply lib/src/package_config_impl.dart]: Would it be easier to read? *boggle* :smiling_imp: 

OK, made it two lines.

File: lib/src/package_config_impl.dart (hunk 2 above)
Comment (lrhn) [reply lib/src/package_config_impl.dart]: That's not a bad idea. Let's throw.

**Pull request merged**